### ISSUE TITLE
Import error

### ISSUE THREAD

**Issue by jgbos** [facebookincubator/submitit#11]
This bug is baffling me, I'm sure this is user error because I normally have no issues with your code.  I have not figured out what is different than my other submissions, but maybe you've seen this before?

```bash
.../python3.8/site-packages/submitit/core/_submit.py", line 7, in <module>
    from .submission import submitit_main
ImportError: attempted relative import with no known parent package
```

**Comment by jgbos** [comment]
Argh, sorry.  This was user error as the job was not configured appropriately.  Not exactly sure why that error happens, but it's not an issue on your end.

**Comment by tullie** [comment]
I'm running into this as well. @jgbos how was your job configured incorrectly? Wondering if i'm doing the same thing.

**Comment by jgbos** [comment]
Sorry, I can't remember exactly what the issue was, wish I would have mentioned it above. I'm sure it was something like setting `slurm_partition` wrong.  Definitely something with the slurm sbatch variables.

**Comment by jrapin** [comment]
This would be weird, sbatch parameters should not interfere with the submission in such a way.
@tullie if you get a small repro I can try to have a look

**Comment by jgbos** [comment]
Honestly I don't remember exactly what caused it, just think it was with those parameters.  Unfortunate I did not document it.

**Comment by tullie** [comment]
So I fixed it by making sure gpus_per_node and ntasks_per_node are always the same. I think this is a pretty common distributed training mistake when using slurm so might be worth having a more relevant error if possible in submitit.

gpus_per_node=cfg.NUM_GPUS,
ntasks_per_node=cfg.NUM_GPUS,

**Comment by jrapin** [comment]
wait, what were your initial settings? I have noooo idea how that could impact the module import :O

**Comment by tullie** [comment]
I have a slurm file:

```
import os
import pathlib
import shutil
import submitit


def copy_and_run_with_config(run_fn, run_config, directory, **cluster_config):
    working_directory = pathlib.Path(directory) / cluster_config["job_name"]
    shutil.copytree(".", working_directory)
    os.chdir(working_directory)
    executor = submitit.SlurmExecutor(folder=working_directory)
    executor.update_parameters(**cluster_config)
    job = executor.submit(run_fn, run_config)
    print(f"job_id: {job}")
```

Then in the calling code I get the error with:

```
    args = parse_args()
     copy_and_run_with_config(
            main,
            args,
            args.working_directory,
            job_name=args.job_name,
            time="72:00:00",
            partition="dev",
            gpus_per_node=args.NUM_GPUS,
            cpus_per_task=10,
            mem="500GB",
            nodes=args.NUM_SHARDS
     )
```

And the error is fixed with:

```
     args = parse_args()
     copy_and_run_with_config(
            main,
            args,
            args.working_directory,
            job_name=args.job_name,
            time="72:00:00",
            partition="dev",
            gpus_per_node=args.NUM_GPUS,
            ntasks_per_node=args.NUM_GPUS,
            cpus_per_task=10,
            mem="500GB",
            nodes=args.NUM_SHARDS
     )
```

So it might be to do with the fact that i'm copying the working directory over and then starting the submitit executor? I don't see how this would affect things but it could be related.

Let me know if you can't reproduce with that @jrapin and i'll link you to the code internally.

**Comment by jrapin** [comment]
> So I fixed it by making sure gpus_per_node and ntasks_per_node are always the same. I think this is a pretty common distributed training mistake when using slurm so might be worth having a more relevant error if possible in submitit.

@tullie I can confirm that `gpus_per_node` and `tasks_per_node` **can** be different (and that it should be allowed, some distributed settings seem to spawn their own tasks as far as I remember). I just tested it with 2 GPUs and 3 tasks, and with 2 GPUS and 1 task, all went fine.

So I would think this is related to copying the working directory, but then again I don't understand why it wouldn't work and then start to work again, did you change something else in the meantime?

And does the changing directory work well? I'm a bit curious, because pickle may link to an absolute path which would revert to the non-copied directory. But I think it only links to a relative path indeed.

**Comment by tullie** [comment]
Fairly confident I didn't change something else in the meantime. I'll try to reproduce again at some point.

I thought something about hidden absolute paths in the pickle file could be related. I haven't had any issues with copying the directory other then potentially this error though.

**Comment by jrapin** [comment]
> I thought something about hidden absolute paths in the pickle file could be related. I haven't had any issues with copying the directory other then potentially this error though.

The one case that is the most flaky is usually submitting a function that is defined in the submission script, and not in a package. Even then I don't think there would be absolute paths, so not sure. And even if there was, I don't understand how that could lead to an import error :s Unless maybe it leads to some sort of circular import, but I have no idea how that could happen :s 

> Fairly confident I didn't change something else in the meantime. I'll try to reproduce again at some point.

Let me know if you have a setup that consistently raise this error, from that point I can investigate, for now I am still completely lost as to what could happen :s

**Comment by gwenzek** [comment]
Could we dump more information when starting submitit ? 
I'd be interested by seeing `PATH` and `PYTHONPATH`.

**Comment by jrapin** [comment]
> Could we dump more information when starting submitit ?

yes if needed, the question would be whether we should always print it? or print it with DEBUG logging?

**Comment by gwenzek** [comment]
Closing, since the bug hasn't been reproduced.
Please reopen if you manage to have more information, or if the bug happen again.

**Comment by jgbos** [comment]
Hi, I'd like to reopen this issue as I have a use case that I can reproduce.  The code below will reproduce the error.  The code mimics how PyTorch Lightning uses their DDP backend.  Lightning launches a subprocess for each additional GPU on a machine.

```python
import sys
import os

import submitit
import subprocess


def cmd():
    if "TEST" not in os.environ:
        print("Initial Process")
    else:
        print("Starting Subprocess")

    if "TEST" not in os.environ:
        env_copy = os.environ.copy()
        env_copy["TEST"] = "1"

        cmd = [sys.executable]
        cmd += sys.argv
        subprocess.Popen(cmd, env=env_copy, cwd=os.getcwd())


def run():
    executor = submitit.AutoExecutor(folder="log_test", cluster="local")
    executor.update_parameters(tasks_per_node=2)
    executor.submit(cmd)


if __name__ == "__main__":
    run()
```

The fix is to change the relative import in `core/_submit.py` to 

```python
from submitit.core.submission import submitit_main
```

**Comment by gwenzek** [comment]
Thanks for the reproducible bug.
I'm surprised by this interaction between submitit and subprocess.
I've already used job starting subprocess between submitit, but here the subprocess is starting submitit itself:
`python submitit/core/_submit.py log_test`

I'm giving me some time to decide if the issue is the relative import, or if we need to change something else,
or if this change can break other jobs.

**Comment by gwenzek** [comment]
I think we have an issue with `python submitit/core/_submit.py log_test` behaving differently from `python -m submitit.core._submit log_test`.

In `submitit` LocalExecutor we start a new process by calling `python -m submitit.core._submit log_test`.
AFAIU Python resolves the module and call `python submitit/core/_submit.py log_test` with correct sys.path
Then `main` starts a new subprocess with `python submitit/core/_submit.py log_test` with a wrong sys.path, and crash.

**Comment by jgbos** [comment]
Interesting, I was unware of this `sys.path` behavior.  I was completely baffled why this error would pop up. This explains errors in the past where I have to use the full path import in files where I use `python -m package.main`.

**Comment by mmuckley** [comment]
I had this issue when I forgot to set `num_nodes` in the PyTorch Lightning `Trainer` for multi-node training, but I did have it set in `submitit`. I think that it causes PyTorch Lightning to fail this check:

https://github.com/PyTorchLightning/pytorch-lightning/blob/ff41d80706c49e34f33215b1f1cf86619b19dfc9/pytorch_lightning/trainer/connectors/slurm_connector.py#L30

and maybe fall back on some other `ddp` mode, looks like PyTorch Elastic but not completely sure...

https://github.com/PyTorchLightning/pytorch-lightning/blob/ff41d80706c49e34f33215b1f1cf86619b19dfc9/pytorch_lightning/accelerators/accelerator_connector.py#L163-L172

It looks like the `subprocess` code from earlier:

https://github.com/PyTorchLightning/pytorch-lightning/blob/09c2020a9325850bc159d2053b30c0bb627e5bbb/tests/utilities/dist.py#L34-L35

I don't know if this is used outside the testing script.

Not sure if this helps at all, just thought I'd relay for others.

**Comment by gwenzek** [comment]
@jgbos @mmuckley Could one of you give me your code calling pytorch lightning ? 
[test/utilities/dist.py](https://github.com/PyTorchLightning/pytorch-lightning/blob/09c2020a9325850bc159d2053b30c0bb627e5bbb/tests/utilities/dist.py#L34-L35) seems like a test utility not the main entry point.
I'm not familiar with torch lightning.
If the function you're calling relies on parsing sys.argv, then we will have problems because sys.argv is changed by submitit.

**Comment by gwenzek** [comment]
I think

**Comment by mmuckley** [comment]
Unfortunately my use case where I encountered this was quite awhile ago so I'm no longer able to reproduce. In summary, my finding was that there was a distributed training configuration mismatch between what I specified for PyTorch Lightning and what I gave to `submitit`, and that rather than anything catching this misconfiguration, PyTorch Lightning launched with an operation mode that led to a `submitit` exception.

This might be reproducible with PyTorch Lightning examples, for example setting the Trainer `num_nodes` to 1 while assigning multiple nodes in `submitit`. In my experience I think an exception still came up with the `submitit.core.submission` import, but it may have been a different one.

**Comment by jgbos** [comment]
Hi @gwenzek apologies for not being responsive.  I need to setup notifications correctly.  I'll see what I can do, I've developed my own lightning backend (or accelerator in lightning terms) that no longer has the subprocess behavior above (a behavior that caused other errors than just this issue).  So I no longer modify sys.argv and depend on submitit to launch all the processes for me :)

Nevertheless, I'll checkout the fix.

**Comment by jgbos** [comment]
@gwenzek your fix is exactly what I changed in submitit to work for my code.  so i can verify that fix is works for me.

**Comment by jgbos** [comment]
Oh, and here's the code in Lightning that can cause the issue:

https://github.com/PyTorchLightning/pytorch-lightning/blob/master/pytorch_lightning/accelerators/ddp_accelerator.py#L141

**Comment by gwenzek** [comment]
The tentative fix is part of [1.2.0](https://pypi.org/project/submitit/1.2.0/) release.
